### PR TITLE
Simple PRE application

### PR DESCRIPTION
**Pull request by vzotova**
**Type of PR:**
- Feature

**Required reviews:** 
- 3

**What this does:**
PRE application with simplified logic (worker configuration)

**Notes for reviewers:**
Based on #2825

### CODE REVIEW

File: nucypher/blockchain/eth/sol/source/contracts/PREApplicationStub.sol [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+    /**
+    * @notice Make a confirmation by worker
+    */
+    function confirmWorkerAddress() external {
+        address operator = _operatorFromWorker[msg.sender];
+        require(isAuthorized(operator), "No stake associated with the worker");
+        OperatorInfo storage info = operatorInfo[operator];
+        require(!info.workerConfirmed, "Worker address is already confirmed");
+        require(msg.sender == tx.origin, " Only worker with real address can make a confirmation");
```
Comment (derekpierre) [inline nucypher/blockchain/eth/sol/source/contracts/PREApplicationStub.sol]:
```suggestion
        require(msg.sender == tx.origin, "Only worker with real address can make a confirmation");
```

**Review by derekpierre — APPROVED** [review]
🎸 

Pretty cool. Nice tests!

File: nucypher/blockchain/eth/sol/source/contracts/PREApplicationStub.sol (hunk 1 above)
Comment (vzotova) [reply nucypher/blockchain/eth/sol/source/contracts/PREApplicationStub.sol]: ✔️

File: nucypher/blockchain/eth/sol/compile/config.py [hunk 2]
```diff
@@ -47,7 +47,7 @@
 LANGUAGE: str = 'Solidity'
 
 # Version of the EVM to compile for. Affects type checking and code generation.
-EVM_VERSION: str = 'berlin'
+EVM_VERSION: str = 'london'
```
Comment (KPrasch) [inline nucypher/blockchain/eth/sol/compile/config.py:50]: 👑

**Review by cygnusv — APPROVED** [review]

File: nucypher/blockchain/eth/sol/source/contracts/SimplePREApplication.sol [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    /**
+    * @notice Get the value of authorized tokens for active operators as well as operators and their authorized tokens
+    * @param _startIndex Start index for looking in operators array
+    * @param _maxOperators Max operators for looking, if set 0 then all will be used
+    * @return allAuthorizedTokens Sum of authorized tokens for active operators
+    * @return activeOperators Array of operators and their authorized tokens. Operators addresses stored as uint256
+    * @dev Note that activeOperators[0] in an array of uint256, but you want addresses. Careful when used directly!
+    */
+    function getActiveOperators(uint256 _startIndex, uint256 _maxOperators)
+        external view returns (uint256 allAuthorizedTokens, uint256[2][] memory activeOperators)
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply nucypher/blockchain/eth/sol/source/contracts/SimplePREApplication.sol:112]: 👍

File: nucypher/blockchain/eth/sol/source/contracts/SimplePREApplication.sol [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    * @param _minWorkerSeconds Min amount of seconds while a worker can't be changed
+    */
+    constructor(
+        IStaking _tStaking,
+        uint256 _minAuthorization,
+        uint256 _minWorkerSeconds
+    ) {
+        require(
+            _tStaking.authorizedStake(address(this), address(this)) == 0,
+            "Wrong input parameters"
```
[the remark this answers is not in the document]
Comment (vzotova) [reply nucypher/blockchain/eth/sol/source/contracts/SimplePREApplication.sol:59]: yep

File: tests/contracts/main/application/test_worker.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+
+from nucypher.blockchain.eth.constants import NULL_ADDRESS
+from eth_utils import to_checksum_address
+
+
+CONFIRMATION_SLOT = 1
+
+
+def test_bond_worker(testerchain, threshold_staking, pre_application, token_economics):
+    creator, operator1, operator2, operator3, operator4, worker1, worker2, worker3, owner3, *everyone_else = \
```
Comment (jMyles) [inline tests/contracts/main/application/test_worker.py:29]: Always loved this line.  Glad to see it getting even more fun.  :-)

File: tests/contracts/contracts/PREApplicationTestSet.sol [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        external
+    {
+        OperatorInfo storage info = operatorInfo[_operator];
+        info.owner = _owner;
+        info.beneficiary = _beneficiary;
+        info.authorizer = _authorizer;
+    }
+
+    function setRoles(address _operator) external {
+        OperatorInfo storage info = operatorInfo[_operator];
```
Comment (jMyles) [inline tests/contracts/contracts/PREApplicationTestSet.sol:60]:
```suggestion
        // If the function is called with only the _operator parameter, we presume that the caller wants that address set for the other roles as well.
        OperatorInfo storage info = operatorInfo[_operator];
```

**Review by jMyles — COMMENTED** [review]
I can't say that I can confirm the values of every local in solidity at every line, but at a high level, this looks entirely reasonable.

In addition to the suggestions and comments, I have a pull request with two minor clarifying (ie, not logical) changes:

https://github.com/vzotova/nucypher/pull/6

File: tests/contracts/contracts/PREApplicationTestSet.sol (hunk 6 above)
Comment (vzotova) [reply tests/contracts/contracts/PREApplicationTestSet.sol:64]: ✔️

**Review by KPrasch — APPROVED** [review]
Approved for merge into the upstream branch `threshold-network`.  @vepkenez and I have incoming changes based over this PR in #2846 and https://github.com/KPrasch/nucypher/pull/30 

Thanks for making this contract available so quickly! It has promoted development of ursula in parallel.

**Pull request merged**